File: solucao_francisco/desafio.dart

void main() {
  
  List<String> SENHAS = ["#forTe1", "senhafraca", "Qu@s1", "Voce@Consegue!2023"];

  for(var senha in senhaForte(SENHAS)){
    print(senha);
  }
  
}

bool minuscula(String palavra){
  for(int i = 0; i < palavra.length; i++){
    if (palavra[i].toLowerCase() == palavra[i]){
      return true;
    }
  }
  return false;
}

bool maiuscula(String palavra){
  for(int i = 0; i < palavra.length; i++){
    if (palavra[i].toUpperCase() == palavra[i]){
      return true;
    }
  }
  return false;
}

bool especiais(String palavra){
  RegExp caracteres = RegExp(r'[^\w\s]');
  if (caracteres.hasMatch(palavra)){
    return true;
  }
  return false;
}

bool numero(String palavra){
  RegExp regExp = RegExp(r'\d');
  return regExp.hasMatch(palavra);
}

List<String> senhaForte(List<String> senhas){
  List<String> fortes = [];
  for (var senha in senhas){
    if(senha.length >= 7 && minuscula(senha) && maiuscula(senha) && especiais(senha) && numero(senha)){
      fortes.add(senha);
    }
  }
  return fortes;
}

File: solucao_francisco/desafio.py
import string


CARACTERES_ESPECIAIS = string.punctuation
SENHAS = ["#forTe1", "senhafraca", "Qu@s1", "Voce@Consegue!2023"]

def maiuscula(palavra: str) -> bool:
    for letra in palavra:
        if letra.isupper():
            return True
    return False

def minuscula(palavra: str) -> bool:
    for letra in palavra:
        if letra.islower():
            return True
    return False

def caracteres_especiais(palavra: str) -> bool:
    for letra in palavra:
        if letra in CARACTERES_ESPECIAIS:
            return True
    return False

def numero(palavra: str) -> bool:
    for letra in palavra:
        if letra.isdigit():
            return True
    return False

def senhas_fortes(senhas: list) -> list:
    fortes = []
    for senha in senhas:
        if len(senha) >= 7 and minuscula(senha) and maiuscula(senha) and caracteres_especiais(senha) and numero(senha):
            fortes.append(senha)
    return fortes

for senha_forte in senhas_fortes(SENHAS):
    print(senha_forte)
        